modules/core/src/main/java/org/apache/ignite/internal/processors/query/aware/IndexRebuildStateStorage.java

```diff
@@ -0,0 +1,412 @@
+    /** Key prefix for the MetaStorage. */
+    public static final String KEY_PREFIX = "rebuild-sql-indexes-";
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** MetaStorage synchronization mutex. */
+    private final Object metaStorageMux = new Object();
+
+    /** Lock. */
+    private final Lock lock = new ReentrantLock(true);
+
+    /**
+     * Stopping the grid.
+     * Guarded by {@link #lock}.
+     */
+    private boolean stopGrid;
+
+    /**
+     * Current states.
+     * Mapping: cache name -> index rebuild state.
+     * Guarded by {@link #lock}.
+     */
+    private final Map<String, IndexRebuildState> states = new HashMap<>();
+
+    /**
+     * Completed states to be removed from the MetaStorage.
+     * Mapping: cache name -> index rebuild state.
+     * Guarded by {@link #lock}.
+     */
+    private final Map<String, IndexRebuildState> toRmv = new HashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param ctx Kernal context.
+     */
+    public IndexRebuildStateStorage(GridKernalContext ctx) {
+        this.ctx = ctx;
+    }
+
+    /**
+     * Callback on start of {@link GridQueryProcessor}.
+     */
+    public void onQueryStart() {
+        ctx.internalSubscriptionProcessor().registerMetastorageListener(this);
+    }
+
+    /**
+     * Callback on start of {@link GridCacheProcessor}.
+     */
+    public void onCacheKernalStart() {
+        lock.lock();
+
+        try {
+            if (!states.isEmpty()) {
+                Set<String> toComplete = new HashSet<>(states.keySet());
+                toComplete.removeAll(ctx.cache().cacheDescriptors().keySet());
+
+                for (String cacheName : toComplete)
+                    completeRebuildIndexes(cacheName);
+            }
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Callback on kernel stop.
+     */
+    public void onKernalStop() {
+        lock.lock();
+
+        try {
+            stopGrid = true;
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Callback on start of rebuild cache indexes.
+     * <p/>
+     * Adding an entry that rebuilding the cache indexes in progress.
+     * If the cache is persistent, then add this entry to the MetaStorage.
+     *
+     * @param cacheCtx Cache context.
+     * @see #onFinishRebuildIndexes
+     */
+    public void onStartRebuildIndexes(GridCacheContext cacheCtx) {
+        lock.lock();
+
+        try {
+            assert !stopGrid;
+
+            IndexRebuildState newState = of(cacheCtx);
+
+            if (!newState.saved())
+                states.putIfAbsent(newState.cacheName(), newState);
+            else {
+                toRmv.remove(newState.cacheName());
+
+                states.compute(newState.cacheName(), (k, preState) -> {
+                    if (preState == null)
+                        return newState;
```

Comment:
u can simplify all this code like :
`states.put(cacheCtx.name(), newState);`
isn`t it ?